Source code should be public by default

lawrence-forooghian opened the pull request:
This replaces https://github.com/dxw/playbook/issues/12.

This is a first draft – I'll sort the writing out once we’ve agreed that the content is fine. I'd like to hear people's thoughts on this and whether there's anything else I need to address.

erbridge commented on rfc-029-repositories-should-be-public-by-default.md:
```diff
@@ -0,0 +1,81 @@
+
+## Problem
+
+- dxw creates a lot of source code repositories – for tools that we use
+  internally, and for client work.
+- When writing new code, we have a choice between making it available to the
+  public and keeping it private for our own use or that of the client.
+- Sometimes the code that we write might be useful to the rest of the world –
+  it solves a common problem, it provides inspiration for how to solve a
+  problem
+- We don’t usually work on projects that require secrecy.
+- We already make most of our repositories public – most notably the Playbook –
+    but until now we haven’t had an official policy, although we’ve [tried
+  to](https://github.com/dxw/playbook/issues/12) in the past. Most of our client
+  projects are also in public repositories.
+- We’ve [already talked
+  about](rfc-018-our-decision-making-process-is-public.md) why we make our RFC
+  decision making process open – the same reasons apply here
```
Nitpick about use `.` here. Can we be consistent, please?

lawrence-forooghian replied:
Yup, I'll sort all that out later – the writing is rubbish at the moment:

> I'll sort the writing out once we’ve agreed that the content is fine.

erbridge replied:
Ah sorry. Missed that until after reviewing.

erbridge commented on rfc-029-repositories-should-be-public-by-default.md:
```diff
@@ -0,0 +1,81 @@
+  about the thing
+- know that they’re investing their energy in something maximally useful and
+  long-lasting
+- can build up a public portfolio of work
```
It also enables us to talk about our work without worrying about NDAs and the like.

lawrence-forooghian replied:
:+1: added

erbridge commented on rfc-029-repositories-should-be-public-by-default.md:
```diff
@@ -0,0 +1,81 @@
+
+- TODO – I can’t think of many reasons why such a repo shouldn’t be public –
+  perhaps it contains secrets e.g. keys? Bring this up at tech team forum
+- TODO – should private repos always have a rationale attached to them
+  explaining why they’re private?
```
I think yes. If we've made it private, I should be able to come to it without that context and understand if it still needs to be private.

lawrence-forooghian replied:
👍 I've added this as a "MUST".

erbridge commented on rfc-029-repositories-should-be-public-by-default.md:
```diff
@@ -0,0 +1,81 @@
+which we would for a client project. For example, we have some repositories
+which are proofs of concept, or constrained by time. If somebody outside of dxw
+sees one of these repositories without understanding the context in which it
+was made, it might give a negative impression of the quality of our work.
```
We could improve our readmes to be clear that things are proofs of concept or works in progress. That would help us work out whether we need to do e.g. security updates for them, too.

lawrence-forooghian replied:
Good idea – added that.

Floppy replied:
Openness is about being honest about your bad code as well as the good stuff. But yes, clearly marking things as work in progress, or "open but not really built for reuse" as a lot of the GDS code was, would be good.

rjw1 replied:
The difference between coding in the open and open source.

jacksonj04 replied:
Is it worth cross-referencing this with the stuff on `CONTRIBUTING.md`, so GitHub surfaces it better for people who just stumble across things?

lawrence-forooghian replied:
Good idea, I'll add a link to the previous RFC.

erbridge commented on rfc-029-repositories-should-be-public-by-default.md:
```diff
@@ -0,0 +1,81 @@
+
+## Next steps
+
+- Start a conversation about how we licence the software in our public
+  repositories, and open an RFC
```
MIT or Public Domain

lawrence-forooghian replied:
I think a bunch of people are going to have a bunch of opinions on this – that's why I've left it as a separate thing 😄

Floppy replied:
![Yeah, well, that's just like, your opinion, man](https://media.giphy.com/media/26BRrSvJUa0crqw4E/giphy.gif)

rjw1 replied:
I was certainly looking forward to the licence debate. (MIT or any BSD derivative would be my choice)

lawrence-forooghian replied:
I've flip-flopped on this one and brought the licence choice into this RFC. When we talked about it as a team everyone seemed pretty chill about MIT.

erbridge commented on rfc-029-repositories-should-be-public-by-default.md:
```diff
@@ -0,0 +1,81 @@
+
+- Start a conversation about how we licence the software in our public
+  repositories, and open an RFC
+- Start a conversation about making GovPress repositories public
```
👍

Floppy commented on rfc-029-repositories-should-be-public-by-default.md:
```diff
@@ -0,0 +1,81 @@
+they directly relate to a client project.
+
+- TODO – I can’t think of many reasons why such a repo shouldn’t be public –
+  perhaps it contains secrets e.g. keys? Bring this up at tech team forum
```
Repos should never contain secrets. Making them public is a good way to avoid the temptation to do this :)

lawrence-forooghian commented:
I‘ve messed something up – there's meant to be some changes after @erbridge’s comments but I think I managed to not push them. I’ll sort it out today.

lawrence-forooghian commented on rfc-029-repositories-should-be-public-by-default.md:
```diff
@@ -0,0 +1,81 @@
+- TODO – should private repos always have a rationale attached to them
+  explaining why they’re private?
+
+When starting a new client project, we SHOULD ask the client whether we can
```
Not all clients are required to, or necessarily interested in, meeting the Service Standard, though.

lawrence-forooghian commented:
I've updated in response to feedback, and added:
- a choice of licence
- the requirement for a `WHY_PRIVATE.md` file for private repos—see explanation in RFC.

Floppy commented on rfc-029-source-code-should-be-open-source-by-default.md:
```diff
@@ -0,0 +1,145 @@
+Standard](https://www.gov.uk/service-manual/service-standard/point-12-make-new-source-code-open).
+
+This proposal doesn’t address things that aren’t hosted on
+GitHub—mainly our GovPress work.
```
Why not? We need to justify, under this, why the govpress sites shouldn't move to github and be open source. It may be just because of time and effort, which is fine, but in that case I'd say remove this line and we will just have to make things conform over time when we can.

Floppy replied:
Ah, I see that's in the "next steps" section. I'd say remove this paragraph if that's the long-term intention, or at least add the word `currently`.

lawrence-forooghian replied:
I think it's because I didn't want to turn it into a conversation about where the GovPress repositories should be hosted. I've removed the sentence and lengthened the "next step".

jacksonj04 commented on rfc-029-source-code-should-be-open-source-by-default.md:
```diff
@@ -0,0 +1,144 @@
+## Summary
+
+We believe in working in the open and sharing our work with others. We
+think that nothing should be secret unless there’s a good reason for
```
One very minor and non-blocking thing - in other RFCs we use straight quotes rather than curly.

lawrence-forooghian replied:
Is that a conscious decision or is it just that people usually write in text editors that don't do smart quotes?

jacksonj04 replied:
I don't think it's a conscious decision, I have no preference either way and am happy to mix, just spotted it and went "that looks unusual".

lawrence-forooghian replied:
👍

erbridge commented on rfc-029-source-code-should-be-open-source-by-default.md:
```diff
@@ -0,0 +1,144 @@
+All new GitHub respositories in the dxw organisation SHOULD be public,
+unless they directly relate to a client project. When these
+repositories are public, they SHOULD contain a copy of the [MIT
+License](https://choosealicense.com/licenses/mit/) in a `LICENSE` file
```
```suggestion
License](https://choosealicense.com/licenses/mit/) in a `LICENCE` file
```

GDS suggests using British English spelling, and it's supported by GitHub and similar.

lawrence-forooghian replied:
👍 changed

erbridge commented on rfc-029-source-code-should-be-open-source-by-default.md:
```diff
@@ -0,0 +1,144 @@
+unless they directly relate to a client project. When these
+repositories are public, they SHOULD contain a copy of the [MIT
+License](https://choosealicense.com/licenses/mit/) in a `LICENSE` file
+in the repository’s root.
```
Who owns the copyright on the licence? `(c) dxw` or do we need a different value?

erbridge commented on rfc-029-source-code-should-be-open-source-by-default.md:
I found this phrasing a little confusing. I kept reading the commas as clause breaks, rather than a list of adjectives.

lawrence-forooghian replied:
Definitely was confusing. I've changed it, thanks.

erbridge commented on rfc-029-source-code-should-be-open-source-by-default.md:
```suggestion
  `LICENCE`
```

As above.

erbridge commented on rfc-029-source-code-should-be-open-source-by-default.md:
👍
